subsystem/src/lib.rs

```diff
@@ -0,0 +1,331 @@
+        let call_rq = CallRequest(action_rx);
+
+        task::spawn(async move {
+            log::info!("Subsystem {}/{} started", appname, name);
+
+            // Perform the subsystem task.
+            subsystem(call_rq, shutdown_rq).await;
+
+            // Signal the intent to shut down to the other parts of the application.
+            let _ = shutting_down_tx.send(()).await;
+
+            // Close the channel to signal the completion of the shutdown.
+            std::mem::drop(shutting_down_tx);
+
+            log::info!("Subsystem {}/{} terminated", appname, name);
+        });
+
+        Subsystem::new(action_tx)
+    }
+
+    /// Start a passive subsystem.
+    ///
+    /// A passive subsystem does not interact with the environment on its own. It only serves calls
+    /// from other subsystems.
+    pub fn start_passive<T: 'static + Send>(&self, name: &'static str, mut obj: T) -> Subsystem<T> {
+        self.start(name, |mut call_rq, mut shutdown_rq| async move {
+            loop {
+                tokio::select! {
+                    () = shutdown_rq.recv() => { break; }
+                    call = call_rq.recv() => { call(&mut obj); }
+                }
+            }
+        })
+    }
+
+    async fn wait_for_subsystems_to_shut_down(mut shutting_down_rx: mpsc::Receiver<()>) {
+        // Wait for the subsystems to go down, signalled by closing the shutting_down channel.
+        while let Some(()) = shutting_down_rx.recv().await {}
+    }
+
+    /// Run the application main task.
+    ///
+    /// Completes when all the subsystems are fully shut down.
+    pub async fn main(mut self) {
+        // The main task just waits for the shutdown signal and coordinates the cleanup at the end.
+        // All other functionality is performed by subsystems.
+
+        // Install the Ctrl-C handler as a subsystem.
+        #[cfg(not(loom))]
+        self.start("ctrl-c", |mut call_rq, mut shutdown_rq| async move {
+            tokio::select! {
+                ctrl_c_signal = tokio::signal::ctrl_c() => {
+                    if ctrl_c_signal.is_err() {
+                        log::info!("Ctrl-C signal handler failed");
+                    }
+                }
+                () = shutdown_rq.recv() => {},
+                call = call_rq.recv() => { call(&mut ()); }
+            };
+        });
+
+        log::info!("Subsystem manager {} running", self.name);
+
+        // Wait for a subsystem to shut down.
+        self.shutting_down_rx.recv().await.expect("All subsystems already dead");
+
+        // Drop the unused shutdown request end
+        std::mem::drop(self.shutdown_request_rx);
+        // Signal the manager is shut down so it does not wait for itself
+        std::mem::drop(self.shutting_down_tx);
+
+        log::info!("Subsytem manager {} shutting down", self.name);
+
+        // Order all the remaining subsystems to shut down.
+        let _ = self.shutdown_request_tx.send(());
```

Comment:
Can't we do this tx/rx of shutdown with a clean future without exposing the internal mechanism?

Reply:
Not sure how this would be done with a plain future. Also I don't see how this exposes internal mechanisms in any significant way. The user does not have direct access to this channel.